tutorial/chapter1.md

```diff
@@ -1,41 +1,219 @@
+```
+
+Create publishers for each image source with an `kpsr::vision_ocv::ImageDataFactory`. `i` is the camera index at the code:
+
+```cpp
+std::vector<kpsr::Publisher<kpsr::vision_ocv::ImageData> *> publishers;
+kpsr::vision_ocv::ImageDataFactory imageCreator([...]);
+    publishers.push_back(
+        eventLoopProvider
+            .getPublisher<kpsr::vision_ocv::ImageData>(prefixChannel + std::to_string(i),
+                                                       0,
+                                                       imageCreator.initializerFunction,
+                                                       imageCreator.eventClonerFunction));
+```
+
+Then create the subscribers to the camera sources and register the post-processing functions. Note the use of `kpsr::vision_ocv::ImageData`.
+
+```cpp
+std::vector<kpsr::Subscriber<kpsr::vision_ocv::ImageData> *> subscribers;
+std::array<int, totalCameras> frameCounters{};
+[...]
+    subscribers.push_back(
+        eventLoopProvider.getSubscriber<kpsr::vision_ocv::ImageData>(channelName));
+    auto writeCameraImageInGrayscale = [i, &frameCounters, outputFolderPrefix](
+                                           const kpsr::vision_ocv::ImageData &event) {
+        cv::Mat outputImage;
+        cv::cvtColor(event.img, outputImage, cv::COLOR_BGR2GRAY);
+        bool statusWrite = cv::imwrite(imgName, outputImage);
+    };
+    subscribers[i]->registerListener(channelName, writeCameraImageInGrayscale);
+```
+
+Create an image service for each image source with `kpsr::vision_ocv::FileImageStreamingService` and `restartIfNoMoreImages=true`:
+
+```cpp
+std::vector<std::unique_ptr<kpsr::vision_ocv::FileImageStreamingService>> readFromDiskServices;
+[...]
+    readFromDiskServices.emplace_back(
+        new kpsr::vision_ocv::FileImageStreamingService([...]
+                                                        publishers[i],
+                                                        inputFolderName,
+                                                        true,
+                                                        width,
+                                                        height));
+    readFromDiskServices[i]->startup();
+```
+
+Run each service independently:
+
+```cpp
+std::vector<std::thread> serviceRunnerThreads;
+[...]
+    serviceRunnerThreads.emplace_back([readTimes, i, &readFromDiskServices]() {
+        for (int j = 0; j < readTimes; j++) {
+            spdlog::info("Service Runner {}. Iteration {}. Thread-id: {}",
+                         i,
+                         j,
+                         getPrintableThreadId());
+            readFromDiskServices[i]->runOnce();
+        }
+```
+
+Finally, clean up and output the `totalFramesListened`:
+
+```cpp
+    serviceRunnerThreads[i].join();
+    readFromDiskServices[i]->shutdown();
+    subscribers[i]->removeListener(prefixChannel + std::to_string(i));
+eventLoopProvider.stop();
+```
+
+Typical output looks like:
```

Comment:
```suggestion
Next is what the typical output of this example looks like. Notice that all listeners belonging to cameras (Lines starting with *Listener*) share the same thread id because they are run in the same *EventLoop*. Service runners are run on different threads and therefore they all have different thread-ids, but they are not the focus of this example.
```